traitsui/examples/demo/Advanced/Dynamic_range_trait_and_editor.py

```diff
@@ -135,13 +135,19 @@ def _get_min_temperature(self):
                 min(int(60.00 / self.fuel_cost), 15))
 
     # Event handlers:
-    @on_trait_change('guests[]')
-    def _guests_modified(self, removed, added):
-        for guest in added:
-            guest.hotel = self
+    @observe('guests.items')
+    def _guests_modified(self, event):
+        # The entire guests list changed
+        if isinstance(event.object, Hotel):
+            for guest in event.new:
+                guest.hotel = self
+        # the contents of the guests list changed
+        else:
+            for guest in event.added:
+                guest.hotel = self
 
     def _add_guest_changed(self):
-        self.guests.append(Guest())
+        self.guests.append(Guest(hotel=self))
```

Comment:
this fixes the previous error, but now it feels like listening to items above is pointless.... As I believe (for this demos sake) the only way `guests.items` will be changed is via the `add_guest` button.  
The cause of this issue is still mysterious to me.  It seems that when the Guest object is instantiated, we have the following trait definitions that need to be sorted out:
```
# The maximum temperature allowed by the guest's plan:
    max_temperature = Property(depends_on='plan')

    # The current room temperature as set by the guest:
    temperature = Range('hotel.min_temperature', 'max_temperature')
```
however, at the time of instantiation the hotel is not defined (when we have the observe decorated listener above, but with on trait change before this worked fine).   

To prevent the error mentioned https://github.com/enthought/traitsui/pull/1519#discussion_r574862947 I simply specify the hotel to be self at the time we try to create the `Guest` object.  But that makes the need to set `guest.hotel = self` above completely redundant...


This all feels somewhat fundamental to the purpose of the demo, but I don't think my understanding of the the dynamic range is entirely there to fully make sense of this.

Reply:
@rahulporuri shall I remove the code for listening to `items` in the above listener now?  Seems strange this example worked before but not with observe without this change on line 150

Reply:
i'm not sure it's worth the effort figuring out what the best way is to structure this example. let's leave it as is for now.